### ISSUE TITLE
Reschedule thread between different core types

### ISSUE THREAD

**Issue by heojun18** [s5z/zsim#191]
Hello, 

I want to ignore stats (cycles, instructions, …) spent inside specific ROI  (not zsim roi)
So, I implement core re-scheduling part based on Per Thread Forwarding #138

// refer to this comment
Benpatsai commented
So that boils down to how to schedule a particular thread from a particular core to another core. One example implementation can be:
1. Implement a magic op to distinguish the main thread from other threads (like register thread)
2. Make the process have multiple core masks, one for the OOO core + the Null core, and one for in-order cores.
3. For non-main thread, use the in-order core mask. For the main thread, use the other mask. You can achieve this by setting the mask vector in the ThreadInfo for a thread.
4. When running into the magic functions, schedule the main thread between those two cores within that mask.

And, here is my code snippet that I implemented.

// my code
VOID HandleMagicOp(THTREADID tid, ADDRINT op) {
…
        Case 1029: // my ROI Begin 
                uint32_t cid = getCid(tid);
                // ARC-HJ schedule this thread to Null Core
                assert(cid != INVALID_CID);
                clearCid(tid);
                zinfo->sched->leave(procIdx, tid, cid);
                SimThreadFini(tid);
                fPtrs[tid] = joinPtrs;
                SimThreadStart(tid);
        Case 1030: // my ROI end
                uint32_t cid = getCid(tid);
                assert(cid != INVALID_CID);
                clearCid(tid);
                zinfo->sched->leave(procIdx, tid, cid);
                SimThreadFini(tid);
                fPtrs[tid] = joinPtrs;
                SimThreadStart(tid);
…
}

And I modified source code (ContextInfo* schedThread(ThreadInfo* th) in scheduler.h and void SimThreadStart(THREADID tid) in sim.cpp) (modify core mask and implement core re-schedulig)

As a result, I succeeded in thread rescheduling (ooo to null or null to ooo in roi region).
But, simulation result changed after rescheduling threads (compare to original code)
        (simulation result is zsim.out stats: phase, cycles, instructions, ...)
All stats are increased (at least 3~4 times, 16 thread program, many roi regions).

In my opinion, when the thread is rescheduled to other type core (ooo <-> null), overhead (that I didn't want) is added (maybe). 
I think that the follow code is relevant to this result.

// my code
        zinfo->sched->leave(procIdx, tid, cid);
        SimThreadFini(tid);
        fPtrs[tid] = joinPtrs;
        SimThreadStart(tid);

In summary, 
My goal: When the thread is simulated in ROI, I want to redirect those stats to NullCore (remove stats in ROI from OOO stats)
Problem: After re-scheduling thread, the simulation result is varied.
So, I want to know how to reschedule thread, without affeting the simulation result (stats)

Thank you !
